Feature: Special date patrol framework + refactoring

larkgz opened the pull request:
* Refactors date stuff to be in a separate module to prevent circular imports
* Switches to use `datetime` module to store dates so that it's easier to do date comparisons (particularly between months)
* Adds game setting to disable special date events
* Add support for special date events that are more than one day long
* Support date exclusive patrols by adding the tag `date_exclusive` and the specified tag for that special date. For Halloween, this is `halloween`.

larkgz committed:
refactor: Move dates to separate module

Circular imports, apparently.

larkgz committed:
feat: Setting for date events

larkgz committed:
feat: Support date ranges

larkgz committed:
feat: Patrol filtering for dates

keyraven commented on scripts/patrol.py:
```diff
@@ -638,6 +639,12 @@ def filter_patrols(self, possible_patrols, biome, patrol_size, current_season, p
                 continue
 
+            # filtering for dates
+            if 'date_exclusive' in patrol.tags:
```
Would it be possible to remove the "date_exclusive" tag, and only check for the special date tags? The "date_exclusive" tag appears redundant.

larkgz committed:
chore: Remove 'date_exclusive' tag

larkgz committed:
perf: Calculate special date once per filter pass

larkgz committed:
fix: special_date filtering logical error

larkgz committed:
Merge remote-tracking branch 'upstream/development' into special-date-patrols

larkgz committed:
fix: Add back special date patrol filtering

Couldn't really do it in the merge so just add it back in a regular commit

larkgz committed:
fix: Correct logical error (real)

Forgot what the purpose of date_exclusive was and didn't replace it properly. Has to check all the tags.
